Chainhooks documentation to run chainhook as a service for Bitcoin and Stacks (#304)
hirosystems/chainhook: docs/how-to-guides/run-chainhook-as-a-service-using-bitcoind.md

---

> title: Run Chainhook as a Service using Bitcoind

Reviewer: I'm curious about these titles. The way the render in the markdown is kind of strange. Do we need these for the website?

Author: I removed the H1 to avoid duplicate titles on the rendered markdown page.

---

# Run Chainhook as a Service using Bitcoind

> This document helps you configure Bitcoind and chainhook nodes to extract on-chain data based on the predicate definition.

Reviewer: ```suggestion
This document helps you configure Bitcoind and Chainhook nodes to extract on-chain data based on the predicate definition.
```

>[!NOTE]
> Make a note of the `rpcuser`, `rpcpassword` and `rpcport` values to use them later in the chainhook configuration.

```conf
datadir=<path-to-your-downloaded-bitcoind>
server=1
rpcuser="bitcoind_username"  # You can set the username here
rpcpassword="bitcoind_password"  #  You can set the password here
rpcport=8332   # You can set your localhost port number here
rpcallowip=0.0.0.0/0
rpcallowip=::/0
txindex=1
listen=1
discover=0
dns=0
dnsseed=0
listenonion=0
rpcserialversion=1
disablewallet=0
fallbackfee=0.00001
rpcthreads=8
blocksonly=1
dbcache=4096

# Start zeromq
zmqpubhashblock=tcp://0.0.0.0:18543
```

Now that you have `bitcoin.conf` file ready with the bitcoind configurations, you can run the bitcoind node.
In the command below, use the path to your `bitcoin.conf` file from your machine and run the command in the terminal.

> [!NOTE]
> The below command is a startup process that might take a few hours to run.

```bash
$ ./bitcoind -conf=<path-to-bitcoin.config>/bitcoin.conf
```
Once the above command is running, you will see `zmq_url` entries in the output, enabling zeromq.

### Configure Chainhook

> In this section, you will configure chainhook to match the network configurations with the bitcoin config file.
> 
> - Install the latest version of chainhook by following [Install Chainhook from source](../getting-started.md#install-chainhook-from-source).

Reviewer: ```suggestion
In this section, you will configure chainhook to match the network configurations with the bitcoin config file. First, [install the latest version of chainhook](../getting-started.md#install-chainhook-from-source).
```

> In this section, you will configure the chainhook and generate a `chainhook.toml` file to connect the chainhook with your bitcoind node. Navigate to the directory where you want to generate `chainhook.toml` file and use the following command in your terminal.

Reviewer: ```suggestion
Next, you will generate a `Chainhook.toml` file to connect Chainhook with your bitcoind node. Navigate to the directory where you want to generate the `Chainhook.toml` file, and use the following command in your terminal:
```

```bash
$ chainhook config generate --testnet
```

> You'll see that `chainhook.toml` file is generated and the file looks like this:

Reviewer: ```suggestion
The following `Chainhook.toml` file should be generated:
```

```toml
[storage]
working_dir = "cache"

# The Http Api allows you to register / deregister
# dynamically predicates.
# Disable by default.
#
[http_api]
http_port = 20456
database_uri = "redis://localhost:6379/"

[network]
mode = "testnet"
bitcoind_rpc_url = "http://localhost:8332" # Must match the rpcport in the bitcoin.conf
bitcoind_rpc_username = "<bitcoind_username>" # Must match the rpcuser in the bitcoin.conf
bitcoind_rpc_password = "<bitcoind_password>" # Must match the rpcpassword in the bitcoin.conf
stacks_node_rpc_url = "http://localhost:20443"

[limits]
max_number_of_bitcoin_predicates = 100
max_number_of_concurrent_bitcoin_scans = 100
max_number_of_stacks_predicates = 10
max_number_of_concurrent_stacks_scans = 10
max_number_of_processing_threads = 16
max_number_of_networking_threads = 16
max_caching_memory_size_mb = 32000

[[event_source]]
tsv_file_url = "https://archive.hiro.so/mainnet/stacks-blockchain-api/mainnet-stacks-blockchain-api-latest"

```

> In the `chainhook.toml` file that gets generated, you'll need to match the network parameters to the `bitcoin.config` that was generated earlier in [this](#prepare-the-bitcoind-node) section to allow chainhook connect to the bitcoin layer.

Reviewer: ```suggestion
In the `Chainhook.toml` file that gets generated, you'll need to match the network parameters to the `bitcoin.config` that was generated earlier in [this](#prepare-the-bitcoind-node) section to allow Chainhook to connect to the bitcoin layer.
```

> Now, in the `chainhook.toml`, update the following network parameters to match the `bitcoin.conf`.
> The Bitcoin node is exposing the rpc endpoints. To protect the endpoints, we are using rpc username and password fields. To run chainhook as a service using Bitcoin, you must match the rpc endpoints username, password, and network ports.
> 
> Make sure the following configurations match before you proceed further in this article.

Reviewer: ```suggestion
The Bitcoin node is exposing the rpc endpoints. To protect the endpoints, we are using rpc username and password fields. To run chainhook as a service using Bitcoin, you must match the rpc endpoints username, password, and network ports.

As such, some configurations will need to match between Chainhook and the bitcoind node. In the `Chainhook.toml`, update the following network parameters to match the `bitcoin.conf`:
```

| bitcoin.conf      | chainhook.toml                |
| -----------       | -----------                   |
| rpcuser           | bitcoind_rpc_username         |
| rpcpassword       | bitcoind_rpc_password         |
| rpcport           | bitcoind_rpc_url              |
| zmqpubhashblock   | bitcoind_zmq_url              |

> In the `chainhook.toml` file,

Reviewer: ```suggestion
In the `Chainhook.toml` file,
```

- Update the `bitcoind_rpc_username` to use the username set for `rpcuser` earlier.
- Update the `bitcoind_rpc_password` to use the password set for  `rpcpassword` earlier.
- Update the `bitcoind_rpc_url` to use the same host and port for the `rpcport` earlier.
- Next, update the `bitcoind_zmq_url` to use the same host and port for the `zmqpubhashblock` that was set earlier.

## Scan blockchain based on predicates

> Now that your bitcoind and chainhook configurations are done, you can scan your blocks by defining your [predicates](../overview.md#if-this-predicate-design). This section helps you with an example JSON file to scan a range of blocks in the blockchain and render the results. To understand the supported predicates for Bitcoin, refer to [how to use chainhook with bitcoin](how-to-use-chainhook-with-bitcoin.md).

Reviewer: ```suggestion
Now that your bitcoind and Chainhook configurations are done, you can scan your blocks by defining your [predicates](../overview.md#if-this-predicate-design). This section helps you with an example JSON file to scan a range of blocks in the blockchain and render the results. To understand the supported predicates for Bitcoin, refer to [how to use chainhook with bitcoin](how-to-use-chainhook-with-bitcoin.md).
```

### Example 1

Run the following command in your terminal to generate a sample JSON file with predicates.

`$ chainhook predicates new ordinals.json --bitcoin`

A JSON file `ordinals.json` is generated.

```json
{
    "uuid": "1",
    "name": "Hello Ordinals",
    "chain": "bitcoin",
    "version": 1,
    "networks": {
        "testnet": {
            "start_block": 777534,
            "end_block": 777540,
            "if_this": {
                "scope": "ordinals_protocol",
                "operation": "inscription_feed"
            },
            "then_that": {
                "file_append": {
                    "path": "inscription_feed.txt"
                }
            }
        }

    }
}
```

> [!NOTE]
> You can get blockchain height and current block by referring to https://explorer.hiro.so/blocks?chain=mainnet

Now, use the following command to scan the blocks based on the predicates defined in the `ordinals.json` file.

``` bash
$ chainhook predicates scan ordinals.json --config-path=./Chainhook.toml
```

The output of the above command will be a text file `inscription_feed.txt` generated based on the predicate definition.

> > [!TIP]
> > To optimize your experience with scanning, the following are a few knobs you can play with:
> > Use of adequate values for `start_block` and `end_block` in predicates will drastically improve the performance.
> > Networking: reducing the number of network hops between the chainhook and the bitcoind processes can also help.

Reviewer: ```suggestion
> [!TIP]
> To optimize your experience with scanning, the following are a few knobs you can play with:
>  - Use of adequate values for `start_block` and `end_block` in predicates will drastically improve the performance.
>  - Reducing the number of network hops between the Chainhook and the bitcoind processes can also help, so your network setup can play a major role in performance.
```

- Set a username of your choice for bitcoind and use it in the `rpcuser` configuration below.
- Set a password of your choice for bitcoind and use it in the `rpcpassword` configuration below.

>[!NOTE]
> Make a note of the `rpcuser`, `rpcpassword` and `rpcport` values to use later in the chainhook configuration.

```conf
datadir=<path-to-your-downloaded-bitcoind>
server=1
rpcuser="bitcoind_username"  --> You can set the username here
rpcpassword="bitcoind_password"  --> You can set the password here
rpcport=8332   --> You can set your localhost port number here
rpcallowip=0.0.0.0/0
rpcallowip=::/0
txindex=1
listen=1
discover=0
dns=0
dnsseed=0
listenonion=0
rpcserialversion=1
disablewallet=0
fallbackfee=0.00001
rpcthreads=8
blocksonly=1
dbcache=4096

# Start zeromq
zmqpubhashblock=tcp://0.0.0.0:18543
```

### Run the bitcoind node

Now that you have `bitcoin.conf` file ready with the bitcoind configurations, you can run the bitcoind node.
In the command below, use the path to your `bitcoin.conf` file from your machine and run the command in the terminal.

```bash

> $ ./bitcoind -conf=<path-to-bitcoin.config>/bitcoin.conf

Reviewer: Since this command is starting a bitcoin node, there's going to be a pretty huge initial sync. It may be worth mentioning that this startup process takes a few hours

```

> Once the above command is running, you will see `zmq_url` entries in the output, ensuring that zeromq is enabled.

Reviewer: I'm not seeing this 🤔 

I think it would be helpful to give an example of what a `zmq_url` entry looks like so users know what to look for

Reviewer: Never mind, I wasn't correctly pointing to the config file! I got this working and see the zmq url
